apps/emqx_durable_storage/src/emqx_ds_lts.erl

```diff
@@ -0,0 +1,619 @@
+                [] -> undefined
+            end
+    end.
+
+-spec trie_insert(trie(), state(), edge()) -> {Updated, state()} when
+    NChildren :: non_neg_integer(),
+    Updated :: false | NChildren.
+trie_insert(Trie, State, Token) ->
+    trie_insert(Trie, State, Token, get_id_for_key(Trie, State, Token)).
+
+%%================================================================================
+%% Internal functions
+%%================================================================================
+
+-spec trie_insert(trie(), state(), edge(), state()) -> {Updated, state()} when
+    NChildren :: non_neg_integer(),
+    Updated :: false | NChildren.
+trie_insert(#trie{trie = Trie, stats = Stats, persist = Persist}, State, Token, NewState) ->
+    Key = {State, Token},
+    Rec = #trans{
+        key = Key,
+        next = NewState
+    },
+    case ets:insert_new(Trie, Rec) of
+        true ->
+            ok = Persist(Key, NewState),
+            Inc =
+                case Token of
+                    ?EOT -> 0;
+                    ?PLUS -> 0;
+                    _ -> 1
+                end,
+            NChildren = ets:update_counter(Stats, State, {2, Inc}, {State, 0}),
+            {NChildren, NewState};
+        false ->
+            [#trans{next = NextState}] = ets:lookup(Trie, Key),
+            {false, NextState}
+    end.
+
+-spec get_id_for_key(trie(), state(), edge()) -> static_key().
+get_id_for_key(#trie{static_key_size = Size}, _State, _Token) ->
+    %% Requirements for the return value:
+    %%
+    %% It should be globally unique for the `{State, Token}` pair. Other
+    %% than that, there's no requirements. The return value doesn't even
+    %% have to be deterministic, since the states are saved in the trie.
+    %%
+    %% The generated value becomes the ID of the topic in the durable
+    %% storage. Its size should be relatively small to reduce the
+    %% overhead of storing messages.
+    %%
+    %% If we want to impress computer science crowd, sorry, I mean to
+    %% minimize storage requirements, we can even employ Huffman coding
+    %% based on the frequency of messages.
+    <<Int:(Size * 8)>> = crypto:strong_rand_bytes(Size),
+    Int.
+
+%% erlfmt-ignore
+-spec do_match_topics(trie(), state(), [binary() | '+'], [binary() | '+' | '#']) ->
+          list().
+do_match_topics(Trie, State, Varying, []) ->
+    case trie_next(Trie, State, ?EOT) of
+        {false, Static} -> [{Static, lists:reverse(Varying)}];
+        undefined -> []
+    end;
+do_match_topics(Trie, State, Varying, ['#']) ->
+    Emanating = emanating(Trie, State, ?PLUS),
+    lists:flatmap(
+        fun
+            ({?EOT, Static}) ->
+                [{Static, lists:reverse(Varying)}];
+            ({?PLUS, NextState}) ->
+                do_match_topics(Trie, NextState, [?PLUS | Varying], ['#']);
+            ({_, NextState}) ->
+                do_match_topics(Trie, NextState, Varying, ['#'])
+        end,
+        Emanating
+    );
```

Comment:
Q: As far as I can understand, with sufficiently populated trie this will result in noticeably large number of matches (and, therefore, streams and iterators) on wide wildcard queries, like `t/#`. Is that correct?

Reply:
Yes. There's not much we can do about clients subscribing to `#`, other than introducing shared subscriptions.